[SPARK-25118][Submit] Persist Driver Logs in Client mode to Hdfs

ankuriitg opened the pull request:
## What changes were proposed in this pull request?
Currently, we do not have a mechanism to collect driver logs if a user chooses
to run their application in client mode. This is a big issue as admin teams need
to create their own mechanisms to capture driver logs.

This commit adds a logger which, if enabled, adds a local log appender to the
root logger and asynchronously syncs it an application specific log file on hdfs
(Spark Driver Log Dir).

Additionally, this collects spark-shell driver logs at INFO level by default.
The change is that instead of setting root logger level to WARN, we will set the
consoleAppender threshold to WARN, in case of spark-shell. This ensures that
only WARN logs are printed on CONSOLE but other log appenders still capture INFO
(or the default log level logs).

## How was this patch tested?
1. Verified that logs are written to local and remote dir
2. Added a unit test case
3. Verified this for spark-shell, client mode and pyspark.
4. Verified in both non-kerberos and kerberos environment
5. Verified with following unexpected termination conditions: Ctrl + C, Driver
OOM, Large Log Files
6. Ran an application in spark-shell and ensured that driver logs were
captured at INFO level
7. Started the application at WARN level, programmatically changed the level to
INFO and ensured that logs on console were printed at INFO level

ankuriitg committed:
[SPARK-25118][Submit] Persist Driver Logs in Yarn Client mode to Hdfs

Currently, we do not have a mechanism to collect driver logs if a user chooses
to run their application in client mode. This is a big issue as admin teams need
to create their own mechanisms to capture driver logs.

This commit adds a logger which, if enabled, adds a local log appender to the
root logger and asynchronously syncs it to hdfs (Spark Driver Log Dir). This log
appender writes the logs to an application specific log file on hdfs and, if
configured, this file is persisted in the Yarn Application Dir on Spark
Application end, allowing for easy discovery/maintenability.

If YARN application dir is not available, or we are not able to write to that
directory, the logs will still be persisted to the configurable Spark Driver Log
Dir (Hdfs).

Additionally, this collects spark-shell driver logs at INFO level by default.
The change is that instead of setting root logger level to WARN, we will set the
consoleAppender threshold to WARN, in case of spark-shell. This ensures that
only WARN logs are printed on CONSOLE but other log appenders still capture INFO
(or the default log level logs).

Testing Done:
1. Verified that logs are written to local/remote dir and then moved to Yarn
Application Dir on application end.
2. Added a unit test case
3. Verified this for spark-shell, client mode and pyspark.
4. Verified in both non-kerberos and kerberos environment
5. Verified with following unexpected termination conditions: Ctrl + C, Driver
OOM, Large Log Files
6. Ran an application at WARN level and ensured that driver logs were
captured at INFO level
7. Started the application at WARN level, programmatically changed the level to
INFO and ensured that logs on console were printed at INFO level

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

SparkQA commented:
**[Test build #96384 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96384/testReport)** for PR 22504 at commit [`d792809`](https://github.com/apache/spark/commit/d792809532ca648d133260c564d33e014a35c135).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/96384/
Test FAILed.

vanzin commented on core/src/main/scala/org/apache/spark/util/logging/DriverLogger.scala:
```diff
@@ -0,0 +1,241 @@
+      val appId = Utils.sanitizeDirName(conf.getAppId)
+      hdfsLogFile = {
+        val rootDir = conf.get(DRIVER_LOG_DFS_DIR.key,
+          "/tmp/driver_logs").split(",").head
```
Fits in previous line.

vanzin commented on core/src/main/scala/org/apache/spark/util/logging/DriverLogger.scala:
```diff
@@ -0,0 +1,241 @@
+      val yarnConf = new YarnConfiguration(hadoopConfiguration)
+      val rootDir = yarnConf.get(YarnConfiguration.NM_REMOTE_APP_LOG_DIR)
+      if (rootDir != null && !rootDir.isEmpty()) {
+        var parentDir = if (UserGroupInformation.getCurrentUser() != null) {
```
`val`

I don't know if `getCurrentUser` will ever be null. But if you're worried I'd just log something and do nothing here. The behavior of your code in that case does not look very appropriate (writing to "/yarn-root-dir/logs/blah") and would probably cause an exception anyway.

vanzin commented on core/src/main/scala/org/apache/spark/util/logging/DriverLogger.scala:
```diff
@@ -0,0 +1,241 @@
+        }
+      }
+    } catch {
+      case nse: NoSuchElementException => logWarning("Couldn't move driver log" +
```
Move code after `=>` to next line. Where is this exception thrown btw? The log message is very specific...

ankuriitg replied:
This exception can be thrown in conf.getAppId. The scenario is if there is an exception before Spark application starts, then this exception will be thrown while trying to close the logger.

vanzin commented on core/src/main/scala/org/apache/spark/util/logging/DriverLogger.scala:
```diff
@@ -0,0 +1,241 @@
+      val rootDir = yarnConf.get(YarnConfiguration.NM_REMOTE_APP_LOG_DIR)
+      if (rootDir != null && !rootDir.isEmpty()) {
+        var parentDir = if (UserGroupInformation.getCurrentUser() != null) {
+          FileUtils.getFile(
```
Will this parent dir already exist at the time this code runs?

I'd be more comfortable with asserting that it does exist, in which case it would have been created by YARN itself, than to automatically create it by calling one of the `FileSystem` methods later in this method...

vanzin commented on core/src/main/scala/org/apache/spark/util/logging/DriverLogger.scala:
```diff
@@ -0,0 +1,241 @@
+        new Path(hdfsLogFile))
+      fileSystem.setPermission(new Path(hdfsLogFile), LOG_FILE_PERMISSIONS)
+      logInfo(
+        s"Moved the local driver log file to spark dfs at: ${hdfsLogFile}")
```
Fits in previous line.

vanzin commented on core/src/main/scala/org/apache/spark/util/logging/DriverLogger.scala:
```diff
@@ -0,0 +1,241 @@
+        // Write all remaining bytes
+        run()
+      } catch {
+        case e: Exception => logError("Failed to write to spark dfs", e)
```
`run()` doesn't throw anything.

vanzin commented on core/src/main/scala/org/apache/spark/util/logging/DriverLogger.scala:
```diff
@@ -0,0 +1,241 @@
+
+  def apply(conf: SparkConf): Option[DriverLogger] = {
+    if ((conf.get(DRIVER_LOG_SYNCTODFS) || conf.get(DRIVER_LOG_SYNCTOYARN))
+      && Utils.isClientMode(conf)) {
```
indent one more level

squito replied:
still looks like the wrong indent here

vanzin commented on core/src/main/scala/org/apache/spark/util/logging/DriverLogger.scala:
```diff
@@ -0,0 +1,241 @@
+    if ((conf.get(DRIVER_LOG_SYNCTODFS) || conf.get(DRIVER_LOG_SYNCTOYARN))
+      && Utils.isClientMode(conf)) {
+      Try[DriverLogger] { new DriverLogger(conf) }
+        .recoverWith { case t: Throwable => logError("Could not add driver logger", t); Failure(t) }
```
Break into multiple lines instead of using `;`

ankuriitg committed:
Minor refactoring

SparkQA commented:
**[Test build #96576 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96576/testReport)** for PR 22504 at commit [`7da9781`](https://github.com/apache/spark/commit/7da9781f1ffe0e93c0ab7e281267bbddf52e0349).

ankuriitg committed:
Minor refactoring: part 2

SparkQA commented:
**[Test build #96577 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96577/testReport)** for PR 22504 at commit [`e13d6c6`](https://github.com/apache/spark/commit/e13d6c6744baafa7e79e5d37fc62d1a2fecd6f36).

SparkQA commented:
**[Test build #96576 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96576/testReport)** for PR 22504 at commit [`7da9781`](https://github.com/apache/spark/commit/7da9781f1ffe0e93c0ab7e281267bbddf52e0349).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/96576/
Test FAILed.

SparkQA commented:
**[Test build #96577 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96577/testReport)** for PR 22504 at commit [`e13d6c6`](https://github.com/apache/spark/commit/e13d6c6744baafa7e79e5d37fc62d1a2fecd6f36).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/96577/
Test FAILed.

bersprockets commented:
retest this please

SparkQA commented:
**[Test build #96717 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96717/testReport)** for PR 22504 at commit [`e13d6c6`](https://github.com/apache/spark/commit/e13d6c6744baafa7e79e5d37fc62d1a2fecd6f36).

SparkQA commented:
**[Test build #96717 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96717/testReport)** for PR 22504 at commit [`e13d6c6`](https://github.com/apache/spark/commit/e13d6c6744baafa7e79e5d37fc62d1a2fecd6f36).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/96717/
Test FAILed.

SparkQA commented:
**[Test build #96729 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96729/testReport)** for PR 22504 at commit [`e13d6c6`](https://github.com/apache/spark/commit/e13d6c6744baafa7e79e5d37fc62d1a2fecd6f36).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/96729/
Test FAILed.

bersprockets commented:
The Py4JJavaError StackOverflow happens pretty reliably. I am guessing its related to the change.

ankuriitg commented:
Yes, I will try to reproduce it locally

ankuriitg commented:
@srowen @jkbradley - it seems that my PR has uncovered some bug in another class. GeneralizedLinearRegressionTrainingSummary.toString actually runs a Spark job, so it ends up into an infinite loop in logDebug() when the log level is DEBUG. Thanks to @vanzin for finding that!

What should be the right change to fix it?

ankuriitg commented:
@srowen - the bug is causing a unit-test failure on this PR. How should I resolve it?

srowen commented:
Just open a separate JIRA, PR to fix it. I suspect we shouldn't run a Spark job in toString.

ankuriitg commented:
Sure, let me do that

tgravescs commented:
while I see this as being a nice feature, I'm a bit hesitant of the yarn logs part of this. You are essentially duplicating the yarn nodemanager logic and assuming you know what its doing.   
If you don't get permissions, format, compression - TFILE format, etc correct then its not going to be readable.

For your use case how are you expecting to get the logs?  just via yarn logs?

Did you look at just using the hadoop log aggregation classes directly?

Putting into some random hdfs directory is obviously much easier the only thing there is user or someone has to make sure they cleanup.

Note I haven't looked at the code in detail yet.

ankuriitg commented:
Thanks for your comment @tgravescs 

Moving the file to yarn logs app dir is optional and users can choose to disable that. In my case, If they only enable syncToHdfs, the file will be moved to the configured hdfs dir.

Regarding cleaning, I have another PR ready, which adds to the existing logic of Spark History Server cleaner and cleans these driver logs as well. It will use the existing max age and polling interval configured for event logs. (I can merge that code in this PR as well)

ankuriitg commented:
Also, for our use case, we are planning to get the logs just via yarn logs

tgravescs commented:
its good its optional, but I think if we say we support it internally to Spark we need to make sure its done properly and works for most yarn setups. 

Did you look at using the hadoop classes at all?  

Just skimming the code it looks like you just move it directly to the hadoop log dir which definitely won't work in our environment.  Does it work in yours?  The file needs to be named node_port, and it  has to be in TFILE format. If that format is not right yarn logs will stop reading once it hits that bad file.

ankuriitg commented:
It worked in my environment but let me look at hadoop classes to see if it can be done in a better manner. Meanwhile, I will merge the cleaner code in this PR.

tgravescs commented:
to clarify what I mean "by work", I mean in order for the "yarn logs" command to read it properly.

ankuriitg commented:
Got it, I didn't test that. Let me verify that and explore alternate options.

ankuriitg committed:
Clean logs from spark driver logs dfs dir, if older than configured max age

This change repurposes the Spark History Server cleaner to delete driver logs
that may be stored in spark driver log dfs dir, if they are older than the
configured max age.

Testing Done:
Changed the polling interval and max age of cleaner and ensured that any
driver logs which are older than configured max age are deleted.

SparkQA commented:
**[Test build #96902 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96902/testReport)** for PR 22504 at commit [`eb751b2`](https://github.com/apache/spark/commit/eb751b293f99e412727cbbd1229dcdd73dfafb54).

SparkQA commented:
**[Test build #96902 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96902/testReport)** for PR 22504 at commit [`eb751b2`](https://github.com/apache/spark/commit/eb751b293f99e412727cbbd1229dcdd73dfafb54).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/96902/
Test FAILed.

ankuriitg committed:
Remove yarn sync

Syncing logs to yarn app dir is difficult as it needs logs to be written by
following certain guidelines in order for it to be read by 'yarn logs' command.
Additionally, those guidelines are internal to Yarn and hence can be changed
without notice, breaking this functionality.

So, this feature will limit itself to just writing the driver logs to
configurable dfs dir.

ankuriitg commented:
I looked further into it and it seems that writing to yarn app dir will mean following certain guidelines: writing filelength, acls, owner, version, using a compression algorithm, etc. And since these guidelines are internal to spark (not exposed via public apis), they can change and break this functionality. So, I have limited the scope of this feature to syncing only to hdfs.

Thanks for bringing it to notice @tgravescs

SparkQA commented:
**[Test build #96956 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96956/testReport)** for PR 22504 at commit [`522fcea`](https://github.com/apache/spark/commit/522fcea207bdccf19b29ffc3ae132da01d254ee0).

ankuriitg committed:
Minor improvement in exception handling

SparkQA commented:
**[Test build #96958 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96958/testReport)** for PR 22504 at commit [`ead6fac`](https://github.com/apache/spark/commit/ead6face05f73b277bf856158e3dffe27a01c58f).

SparkQA commented:
**[Test build #96956 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96956/testReport)** for PR 22504 at commit [`522fcea`](https://github.com/apache/spark/commit/522fcea207bdccf19b29ffc3ae132da01d254ee0).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/96956/
Test PASSed.

SparkQA commented:
**[Test build #96958 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96958/testReport)** for PR 22504 at commit [`ead6fac`](https://github.com/apache/spark/commit/ead6face05f73b277bf856158e3dffe27a01c58f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/96958/
Test PASSed.

NiharS commented on core/src/main/scala/org/apache/spark/internal/Logging.scala:
```diff
@@ -192,7 +211,15 @@ private[spark] object Logging {
-        LogManager.getRootLogger().setLevel(defaultRootLevel)
+        val rootLogger = LogManager.getRootLogger()
+        rootLogger.setLevel(defaultRootLevel)
+        rootLogger.getAllAppenders().asScala.foreach { tmp =>
```
Any reason not to just iterate through `consoleAppenderToThreshold.keys()`? Not a huge deal but cuts down on a bit of work

ankuriitg replied:
Not really, consoleAppenderToThreshold was added later. Additionally, currently I do not store any consoleAppender which have null threshold into this map. In order for doing this way, I will need to store those as well. Dropping this, as in most scenarios, this will require more storage and minor performance improvement.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -806,6 +806,22 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
+      }
+    } catch {
+      case nse: NoSuchElementException => // no-op
+      case t: Throwable => logError("Failed to delete driver logs", t)
```
Is there a good reason to catch `Throwable` instead of `Exception`?

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -806,6 +806,22 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
+        }
+      }
+    } catch {
+      case nse: NoSuchElementException => // no-op
```
What call throws this exception?

ankuriitg replied:
conf.get("spark.driver.log.dfsDir")

vanzin replied:
It's an optional config. If you use the config constant this will go away. (And even in the legacy case you should use `conf.getOption` in those cases.)

vanzin commented on core/src/main/scala/org/apache/spark/internal/Logging.scala:
```diff
@@ -143,13 +147,27 @@ trait Logging {
           }
-          rootLogger.setLevel(replLevel)
+          rootLogger.getAllAppenders().asScala.foreach { tmp =>
+            tmp match {
```
You can match directly in the body of the `foreach`, no need for the extra level.

vanzin commented on core/src/main/scala/org/apache/spark/internal/Logging.scala:
```diff
@@ -166,6 +184,7 @@ private[spark] object Logging {
   @volatile private var initialized = false
   @volatile private var defaultRootLevel: Level = null
   @volatile private var defaultSparkLog4jConfig = false
+  val consoleAppenderToThreshold = new ConcurrentHashMap[ConsoleAppender, Priority]()
```
`private`?

vanzin commented on core/src/main/scala/org/apache/spark/internal/config/package.scala:
```diff
@@ -48,6 +48,19 @@ package object config {
+      .stringConf
+      .createWithDefault("%d{yy/MM/dd HH:mm:ss.SSS} %t %p %c{1}: %m%n")
+
+  private[spark] val DRIVER_LOG_SYNCTODFS =
```
With the removal of the YARN code, does it even make sense to have this option? Seems like setting `DRIVER_LOG_DFS_DIR` would be enough to enable this.

Only advantage of it is that it makes it simpler to disable the logger if the default config file has it enabled.

ankuriitg replied:
Yes, as you said I prefer to keep the functionality of disabling/enabling separate from log dir. But I do see that removing this reduces the number of configurations.

I am ok either ways. Please let me know if I should remove this.

vanzin replied:
Let's leave it. But there's some usage of the enabled flag in your code that you can still clean up.

vanzin commented on core/src/main/scala/org/apache/spark/util/logging/DriverLogger.scala:
```diff
@@ -0,0 +1,175 @@
+  private var localLogFile: String = FileUtils.getFile(
+    Utils.getLocalDir(conf), "driver_logs", DRIVER_LOG_FILE).getAbsolutePath()
+  private var writer: Option[HdfsAsyncWriter] = None
+  private val syncToDfs: Boolean = conf.get(DRIVER_LOG_SYNCTODFS)
```
This doesn't make sense anymore since you do not instantiate the logger when it's false.

vanzin commented on core/src/main/scala/org/apache/spark/util/logging/DriverLogger.scala:
```diff
@@ -0,0 +1,175 @@
+
+  private def addLogAppender(): Unit = {
+    val appenders = LogManager.getRootLogger().getAllAppenders()
+    val layout = if (appenders.hasMoreElements()) {
```
I don't think this logic is right. It should be:

- if the user has defined the config, use it
- otherwise default to the first appender's layout
- if that is not set, then use a hardcoded default

The first two are inverted in your case. You could achieve that with `.createWithDefaultFunction` in the config builder. But really I'd just make it an optional config, and here do:

- if the user has defined the config, use it
- otherwise default to the first appender's layout
- if that is not set, then use `new PatternLayout()`.

Which I think is what I suggested before.

vanzin commented on core/src/main/scala/org/apache/spark/util/logging/DriverLogger.scala:
```diff
@@ -0,0 +1,175 @@
+  private val DRIVER_LOG_FILE = "driver.log"
+  private val LOG_FILE_PERMISSIONS = new FsPermission(Integer.parseInt("770", 8).toShort)
+
+  private var localLogFile: String = FileUtils.getFile(
```
Slight preference to use:

```
new File(Utils.createTempDir(root = Utils.getLocalDir(conf), namePrefix = "driver_log"), DRIVER_LOG_FILE)
```

That also means Spark takes care of deleting the directory for you later.

ankuriitg replied:
Utils.getLocalDir also does that actually. Spark deletes the directory that it creates.

vanzin commented on core/src/main/scala/org/apache/spark/util/logging/DriverLogger.scala:
```diff
@@ -0,0 +1,175 @@
+      case e: Exception =>
+        logError(s"Error in persisting driver logs", e)
+    } finally {
+      try {
```
This is not needed if you follow my previous suggestion, but there's `Utils.tryLogNonFatalError`.

vanzin commented on core/src/main/scala/org/apache/spark/util/logging/DriverLogger.scala:
```diff
@@ -0,0 +1,175 @@
+    }
+    private var fileSystem: FileSystem = FileSystem.get(hadoopConf)
+    private val outputStream: FSDataOutputStream = fileSystem.create(new Path(hdfsLogFile), true)
+    fileSystem.setPermission(new Path(hdfsLogFile), LOG_FILE_PERMISSIONS)
```
If this throws you'll leave the stream opened.

vanzin commented on core/src/main/scala/org/apache/spark/util/logging/DriverLogger.scala:
```diff
@@ -0,0 +1,175 @@
+    }
+
+    def close(): Unit = {
+      try {
```
Return early if `streamClosed`?

ankuriitg replied:
This should not be called twice, so it is not needed.

vanzin replied:
It's good practice to write idempotent close methods. "Should not be called twice" is part of the famous last words list.

vanzin commented on core/src/main/scala/org/apache/spark/util/logging/DriverLogger.scala:
```diff
@@ -0,0 +1,175 @@
+    def closeWriter(): Unit = {
+      try {
+        threadpool.execute(new Runnable() {
+          def run(): Unit = HdfsAsyncWriter.this.close()
```
override

vanzin commented on core/src/main/scala/org/apache/spark/util/logging/DriverLogger.scala:
```diff
@@ -0,0 +1,175 @@
+    }
+  }
+
+  private class HdfsAsyncWriter(appId: String, hadoopConf: Configuration) extends Runnable
```
s/HDFS/DFS in these names.

vanzin commented on core/src/test/scala/org/apache/spark/util/logging/DriverLoggerSuite.scala:
```diff
@@ -0,0 +1,109 @@
+
+  test("driver logs are persisted") {
+    val sc = getSparkContext()
+    // Wait for application to start
```
What exactly is this waiting for? Sleeping is generally the wrong way to wait for something.

You should also extend `LocalSparkContext`.

If you really need to wait, you should do it in `getSparkContext()` to avoid having to add the wait logic to every test.

vanzin commented on core/src/main/scala/org/apache/spark/util/logging/DriverLogger.scala:
```diff
@@ -0,0 +1,175 @@
+    }
+  }
+
+  def stop(): Unit = {
```
Don't you need to remove the appender you added earlier?

Otherwise one `SparkContext` may mess up other contexts started later. e.g. you'll get two appenders trying to write to the same file with your current code, or you'll have the appender writing stuff for a context that did not request it.

vanzin commented on core/src/test/scala/org/apache/spark/util/logging/DriverLoggerSuite.scala:
```diff
@@ -0,0 +1,109 @@
+    sc.stop()
+    // On application end, file is moved to Hdfs (which is a local dir for this test)
+    assert(!driverLogsDir.exists())
+    val hdfsDir = FileUtils.getFile("/tmp/hdfs_logs/", app_id)
```
You should create a constant for `/tmp/hdfs_logs/` or read it from the SparkConf object.

vanzin commented on core/src/test/scala/org/apache/spark/util/logging/DriverLoggerSuite.scala:
```diff
@@ -0,0 +1,109 @@
+    }
+
+    // After 5 secs, file contents are synced to Hdfs (which is a local dir for this test)
+    Thread.sleep(6000)
```
I'm not a fan of code that relies on exact timing between threads for things to work.

This would be better if you drove the `HdfsAsyncWriter` class manually from the test, instead of indirectly through a `SparkContext`. The you can also control the log file flushing explicitly instead of the hack you have above.

vanzin commented:
Oh, you should also remove yarn from the PR title. This isn't really related to yarn in any way (anymore).

ankuriitg committed:
Incorporating review comments

1. Minor refactoring
2. Added information about driver logs to configuration.md
3. Improved exception handling
4. Minor refactoring of test case

ankuriitg commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -806,6 +806,22 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
+    try {
+      val hdfsDir = conf.get("spark.driver.log.dfsDir")
+      val appDirs = fs.listLocatedStatus(new Path(hdfsDir))
+      while (appDirs.hasNext()) {
```
Didn't work. Maybe because it is a RemoteIterator?

vanzin replied:
Could be. If it doesn't work this is fine.

SparkQA commented:
**[Test build #97177 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97177/testReport)** for PR 22504 at commit [`3b9fc9f`](https://github.com/apache/spark/commit/3b9fc9f9beeb31c701850c431755be17bf5c1648).

ankuriitg commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -806,6 +806,22 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
+        val appDirStatus = appDirs.next()
+        if (appDirStatus.getModificationTime() < maxTime) {
+          logInfo(s"Deleting expired driver log for: ${appDirStatus.getPath().getName()}")
+          deleteLog(appDirStatus.getPath())
```
I added the documentation but I am still relying on the existing history server configurations. This is because I am using an existing cleaner. Otherwise, I may have to create another cleaner for this, which can run at a different interval and has a different configured max age.

Please let me know if I should create another cleaner for this or just provide a configuration for a different max age.

vanzin replied:
It's pretty trivial to create a different task to run this code you're adding...

SparkQA commented:
**[Test build #97177 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97177/testReport)** for PR 22504 at commit [`3b9fc9f`](https://github.com/apache/spark/commit/3b9fc9f9beeb31c701850c431755be17bf5c1648).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/97177/
Test PASSed.

ankuriitg committed:
Incorporating review comments: Part 2

1. Minor refactoring
2. Added new configurations for driver log cleaner in spark history server
3. Made changes to driver log cleaner code to utilize those configurations
4. Updated documentation in configuration.md

ankuriitg committed:
Change test in DriverLoggerSuite such that it does not depend on Thread.sleep()

SparkQA commented:
**[Test build #97211 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97211/testReport)** for PR 22504 at commit [`2bbf137`](https://github.com/apache/spark/commit/2bbf1370ab72a0183cc3ccdc04851b2b6dadf5dc).

SparkQA commented:
**[Test build #97210 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97210/testReport)** for PR 22504 at commit [`7fb94fd`](https://github.com/apache/spark/commit/7fb94fd5273f19d91ec74584d2efa6b33ce371c4).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/97210/
Test PASSed.

SparkQA commented:
**[Test build #97211 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97211/testReport)** for PR 22504 at commit [`2bbf137`](https://github.com/apache/spark/commit/2bbf1370ab72a0183cc3ccdc04851b2b6dadf5dc).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/97211/
Test PASSed.

vanzin commented on core/src/test/scala/org/apache/spark/util/logging/DriverLoggerSuite.scala:
```diff
@@ -0,0 +1,116 @@
+
+  private def getSparkContext(): SparkContext = {
+    val conf = new SparkConf()
+    conf.set("spark.local.dir", "/tmp")
```
Don't do this. Tests run with a pre-configured temp directory, and should not write to `/tmp`.

If you need to create a temp directory, do so explicitly using the existing `Utils` API.

ankuriitg committed:
Review Comments: Minor refactoring

SparkQA commented:
**[Test build #97414 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97414/testReport)** for PR 22504 at commit [`a46734c`](https://github.com/apache/spark/commit/a46734c46355b31863178ff9f0169118ca15a695).

ankuriitg commented on core/src/test/scala/org/apache/spark/util/logging/DriverLoggerSuite.scala:
```diff
@@ -0,0 +1,116 @@
+  }
+
+  test("driver logs are synced to hdfs continuously") {
+    val sc = getSparkContext()
```
I changed the test to just check the final output, so it does not rely on loop logging anymore. That way, I don't need to expose `SparkContext.driverLogger` as well

SparkQA commented:
**[Test build #97414 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97414/testReport)** for PR 22504 at commit [`a46734c`](https://github.com/apache/spark/commit/a46734c46355b31863178ff9f0169118ca15a695).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/97414/
Test PASSed.

squito commented on core/src/main/scala/org/apache/spark/util/logging/DriverLogger.scala:
```diff
@@ -0,0 +1,196 @@
+
+  // Visible for testing
+  private[spark] class DfsAsyncWriter(appId: String, hadoopConf: Configuration) extends Runnable
+    with Logging {
```
nit: double indent

squito commented on core/src/main/scala/org/apache/spark/util/logging/DriverLogger.scala:
```diff
@@ -0,0 +1,196 @@
+        }
+        outputStream.hflush()
+      } catch {
+        case e: Exception => logError("Failed to write to spark dfs", e)
```
its not really "spark dfs".  Maybe "Failed copying driver logs to dfs".

If there is a consistent error, you're going to see this error msg a lot as you'll keep calling `run()`.  I'm not sure how common that will be and if we should do some special handling (eg. only include the full stack trace the first time or something like that).

ankuriitg replied:
There can be cases when there are consistent error - such that someone deleted the file from filesystem, changed the permission, etc. But mostly these will be one off, so I am leaving the exception handling as it is for now. We can certainly change this later on if we discover otherwise.

squito commented on core/src/main/scala/org/apache/spark/util/logging/DriverLogger.scala:
```diff
@@ -0,0 +1,196 @@
+          case t: Throwable =>
+            logError("Could not add driver logger", t)
+            Failure(t)
+        }.toOption
```
probably simpler as a regular `try / catch`

```scala
try {
  Some(new DriverLogger(conf))
}  catch {
  case t: Throwable =>
    logError("Could not add driver logger", t)
    None
}
```

ankuriitg replied:
I will leave it as it is for now unless others also feel the same way, as I came to this implementation based on review comments.

squito commented on core/src/test/scala/org/apache/spark/util/logging/DriverLoggerSuite.scala:
```diff
@@ -0,0 +1,83 @@
+
+class DriverLoggerSuite extends SparkFunSuite with LocalSparkContext {
+
+  private val DRIVER_LOG_DIR_DEFAULT = "/tmp/hdfs_logs"
```
use a relative dir, not an absolute one, for testing data.  (what if a developer happened to have real data at /tmp/hdfs_logs before running this?)

squito commented on core/src/test/scala/org/apache/spark/util/logging/DriverLoggerSuite.scala:
```diff
@@ -0,0 +1,83 @@
+    assert(files(0).getName.equals("driver.log"))
+
+    sc.stop()
+    // On application end, file is moved to Hdfs (which is a local dir for this test)
```
comment is a bit misleading (or maybe out of date) -- its synced continuously, right?

squito commented on docs/configuration.md:
```diff
@@ -266,6 +266,37 @@ of the most common options to set are:
+  <td><code>spark.driver.log.syncToDfs.enabled</code></td>
+  <td>false</td>
+  <td>
+    If true, spark application running in client mode will sync driver logs to a persistent storage, configured
```
it seems like there is a mismatch between what is described here and what is implemented.  Do you intend to support a configuration where you are only logging to local disk, and not syncing to dfs?  It doesn't seem like `DriverLogger.apply` allows that.  And if you do intend to support that, and want it controlled by this configuration, then I'd remove the "syncToDfs" from this name.

ankuriitg replied:
I guess, what I meant to say here is that this feature enables syncing to dfs. We do not support syncing to local disk and that is just a implementation detail, not a feature.

I am good with using any alternate wording, if that will represent the intent better.

squito replied:
I guess I just don't like using the word "sync" here, it makes me think the logs are getting stored somewhere, just not synced to persistent storage, even if this is false.

How about renaming the conf to "spark.driver.log.persistToDfs.enabled" and rewording this to

If true, spark application running in client mode will write driver logs to a persistent storage, configured in spark.driver.log.dfsDir.  If spark.driver.log.dfsDir is not configured, driver logs will not be stored to persistent storage.  Additionally, enable the cleaner by setting spark.history.fs.driverlog.cleaner.enabled to true.

?

ankuriitg replied:
Makes sense. Updated the config name and the wording.

ankuriitg committed:
Incorporating review comments: Part 3

Minor Refactoring

squito commented on core/src/main/scala/org/apache/spark/util/logging/DriverLogger.scala:
```diff
@@ -0,0 +1,196 @@
+
+  def apply(conf: SparkConf): Option[DriverLogger] = {
+    if (conf.get(DRIVER_LOG_SYNCTODFS)
+        && conf.get(DRIVER_LOG_DFS_DIR).isDefined
```
do you want to at least warn the user if sync_to_dfs is enabled but log_dfs_dir is undefined?  Or is there a useful scenario when you'd want that?

ankuriitg replied:
Sure, added a warning message in that case

ankuriitg committed:
Incorporating review feedback: Part 4

Minor refactoring

SparkQA commented:
**[Test build #97560 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97560/testReport)** for PR 22504 at commit [`07d2df8`](https://github.com/apache/spark/commit/07d2df87fddd540637b054b643eb5484c5e58eaf).

vanzin commented on core/src/main/scala/org/apache/spark/util/logging/DriverLogger.scala:
```diff
@@ -0,0 +1,202 @@
+  def apply(conf: SparkConf): Option[DriverLogger] = {
+    if (conf.get(DRIVER_LOG_PERSISTTODFS) && Utils.isClientMode(conf)) {
+      if (conf.get(DRIVER_LOG_DFS_DIR).isDefined) {
+        Try[DriverLogger] {
```
This is actually simpler with a try..catch. There might have been a reason for this before but in it's current shape, a try..catch is simpler and actually easier to understand.

vanzin commented on core/src/test/scala/org/apache/spark/util/logging/DriverLoggerSuite.scala:
```diff
@@ -0,0 +1,83 @@
+
+class DriverLoggerSuite extends SparkFunSuite with LocalSparkContext {
+
+  private var rootHdfsDir : File = _
```
s/hdfs/dfs in all this file.

vanzin commented on core/src/main/scala/org/apache/spark/util/logging/DriverLogger.scala:
```diff
@@ -0,0 +1,202 @@
+  private val LOG_FILE_PERMISSIONS = new FsPermission(Integer.parseInt("770", 8).toShort)
+
+  private var localLogFile: String = FileUtils.getFile(
+    Utils.getLocalDir(conf), "driver_logs", DRIVER_LOG_FILE).getAbsolutePath()
```
I'd really be more comfortable with using a temp dir rather than a hardcoded directory name...

Or at least use `__driver_logs__` since that syntax conventionally means things that user applications should not use.

And use a constant so you don't have to hardcode the value in the test.

vanzin commented on docs/configuration.md:
```diff
@@ -266,6 +266,37 @@ of the most common options to set are:
+    Base directory in which Spark driver logs are synced, if spark.driver.log.syncToDfs.enabled is true.
+    Within this base directory, Spark creates a sub-directory for each application, and logs the driver logs
+    specific to the application in this directory. Users may want to set this to a unified location like an
+    HDFS directory so driver log files can be persisted for later usage. This directory should allow any spark
```
Spark

vanzin commented on docs/configuration.md:
```diff
@@ -266,6 +266,37 @@ of the most common options to set are:
+    Within this base directory, Spark creates a sub-directory for each application, and logs the driver logs
+    specific to the application in this directory. Users may want to set this to a unified location like an
+    HDFS directory so driver log files can be persisted for later usage. This directory should allow any spark
+    user to read/write files and the spark history server user to delete files. Additionally, older logs from
```
Spark History Server

vanzin commented on docs/configuration.md:
```diff
@@ -266,6 +266,37 @@ of the most common options to set are:
+    specific to the application in this directory. Users may want to set this to a unified location like an
+    HDFS directory so driver log files can be persisted for later usage. This directory should allow any spark
+    user to read/write files and the spark history server user to delete files. Additionally, older logs from
+    this directory are cleaned by Spark History Server if spark.history.fs.driverlog.cleaner.enabled is true.
```
Put config names between `<code></code>`.

This is also wrong, since you do not have an explanation of this option in the documentation. It will be true also if you just enable the normal SHS cleaner. You should explain that option explicitly and say its default value is the value of the other config.

The SHS options are in monitoring.md.

ankuriitg replied:
I provided the information that driverlog cleaner uses a fallback option in its own documentation.

vanzin replied:
That documentation is not user visible.

vanzin commented on docs/configuration.md:
```diff
@@ -266,6 +266,37 @@ of the most common options to set are:
+  <td><code>spark.driver.log.dfsDir</code></td>
+  <td>(none)</td>
+  <td>
+    Base directory in which Spark driver logs are synced, if spark.driver.log.syncToDfs.enabled is true.
```
spark.driver.log.syncToDfs.enabled  does not exist.

ankuriitg committed:
Minor config name change

vanzin commented on core/src/test/scala/org/apache/spark/util/logging/DriverLoggerSuite.scala:
```diff
@@ -0,0 +1,83 @@
+  private var rootHdfsDir : File = _
+
+  override def beforeAll(): Unit = {
+    rootHdfsDir = Utils.createTempDir(namePrefix = "hdfs_logs")
```
It's good practice to call `super.beforeAll` and `super.afterAll`. But here you don't need these; `createTempDir` cleans things up automatically for you.

SparkQA commented:
**[Test build #97561 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97561/testReport)** for PR 22504 at commit [`c6c0432`](https://github.com/apache/spark/commit/c6c04328b2072583fbd213bcefa05fb2d4b82965).

vanzin commented on core/src/test/scala/org/apache/spark/util/logging/DriverLoggerSuite.scala:
```diff
@@ -0,0 +1,83 @@
+    val driverLogFile = hdfsFiles.filter(f => f.getName.equals("driver.log")).head
+    val hdfsIS = new BufferedInputStream(new FileInputStream(driverLogFile))
+    assert(hdfsIS.available() > 0)
+    JavaUtils.deleteRecursively(hdfsDir)
```
This line and the next are unnecessary.

vanzin commented on core/src/test/scala/org/apache/spark/util/logging/DriverLoggerSuite.scala:
```diff
@@ -0,0 +1,83 @@
+    assert(files(0).getName.equals("driver.log"))
+
+    sc.stop()
+    // File is continuously synced to Hdfs (which is a local dir for this test)
```
Comment is unrelated to any of the code around it.

vanzin commented on core/src/test/scala/org/apache/spark/util/logging/DriverLoggerSuite.scala:
```diff
@@ -0,0 +1,83 @@
+    assert(!driverLogsDir.exists())
+    val hdfsDir = FileUtils.getFile(sc.getConf.get(DRIVER_LOG_DFS_DIR).get, app_id)
+    assert(hdfsDir.exists())
+    val hdfsFiles = hdfsDir.listFiles()
```
All the checks below are basically:

```
hdfsFiles.exists { f => f.getName() == DRIVER_LOG_FILE && f.length() > 0 }
```

vanzin commented on core/src/main/scala/org/apache/spark/util/logging/DriverLogger.scala:
```diff
@@ -0,0 +1,202 @@
+      with Logging {
+
+    private var streamClosed = false
+    private val fileSystem: FileSystem = FileSystem.get(hadoopConf)
```
See my previous comment for why this call is not right.

ankuriitg committed:
Incorporating review feedback: Part 5

1. Minor refactoring
2. Added documentation for new Spark History Server configs
3. Ensured that driver logs being written to are not cleaned up, by using file length

SparkQA commented:
**[Test build #97565 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97565/testReport)** for PR 22504 at commit [`a0a85b3`](https://github.com/apache/spark/commit/a0a85b3eff0f115b983ce5ba3214e09f8ee90dd2).

ankuriitg commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -800,14 +817,33 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
+        val appDirStatus = appDirs.next()
+        if (appDirStatus.getModificationTime() < maxTime) {
+          logInfo(s"Deleting expired driver log for: ${appDirStatus.getPath().getName()}")
+          deleteLog(driverLogFs.get, appDirStatus.getPath())
```
Thanks for this. I did not realize that `modtime` is not updated always. Added a fix. The fix will clean the logs when they are older than max(2 * cleaner_interval, max_age).

vanzin replied:
Isn't that just delaying the problem I described?

If you look at how this code handles that for event logs, it's way more complicated than that.

ankuriitg replied:
Sorry, I don't understand your comment. Can you elaborate please?

The problem that you described was the old cleaner logic may delete (or try to delete) a log for an active app. With the solution that I added, I am tracking the fileLength and lastProcessedTime. If the fileLength is updated, I do not delete the file and update the lastProcessedTime instead. And I rely on lastProcessedTime to determine whether it is time to delete the file.

Please let me know if there is an issue in the logic.

vanzin replied:
So you *are* doing what the current code does (I haven't reviewed the code yet). Your comment made it sound like you were just waiting longer to delete the log.

ankuriitg replied:
Ohh right, the fix is similar to what exists currently. I just wanted to add the note that logs will be deleted after that time period.

ankuriitg commented on core/src/main/scala/org/apache/spark/util/logging/DriverLogger.scala:
```diff
@@ -0,0 +1,196 @@
+      case e: Exception =>
+        logError(s"Error in persisting driver logs", e)
+    } finally {
+      Utils.tryLogNonFatalError(JavaUtils.deleteRecursively(
```
I did not understand this comment.

vanzin replied:
```
Utils.tryLogNonFatalError {
   // code goes here
}
```

SparkQA commented:
**[Test build #97566 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97566/testReport)** for PR 22504 at commit [`65c0800`](https://github.com/apache/spark/commit/65c080032bbea82024f1ab14cb43c771f6157fc4).

SparkQA commented:
**[Test build #97557 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97557/testReport)** for PR 22504 at commit [`18b823f`](https://github.com/apache/spark/commit/18b823fd38d93d45c9121b9802bb9206755265bd).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
**[Test build #97561 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97561/testReport)** for PR 22504 at commit [`c6c0432`](https://github.com/apache/spark/commit/c6c04328b2072583fbd213bcefa05fb2d4b82965).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/97561/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/97557/
Test PASSed.

ankuriitg committed:
Updated config documentation

SparkQA commented:
**[Test build #97575 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97575/testReport)** for PR 22504 at commit [`6e14137`](https://github.com/apache/spark/commit/6e141373910df8c159429c367a8e937ad71eccbe).

SparkQA commented:
**[Test build #97566 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97566/testReport)** for PR 22504 at commit [`65c0800`](https://github.com/apache/spark/commit/65c080032bbea82024f1ab14cb43c771f6157fc4).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/97566/
Test FAILed.

SparkQA commented:
**[Test build #97560 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97560/testReport)** for PR 22504 at commit [`07d2df8`](https://github.com/apache/spark/commit/07d2df87fddd540637b054b643eb5484c5e58eaf).
 * This patch **fails from timeout after a configured wait of `400m`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/97560/
Test FAILed.

SparkQA commented:
**[Test build #97565 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97565/testReport)** for PR 22504 at commit [`a0a85b3`](https://github.com/apache/spark/commit/a0a85b3eff0f115b983ce5ba3214e09f8ee90dd2).
 * This patch **fails PySpark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/97565/
Test FAILed.

SparkQA commented:
**[Test build #97575 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97575/testReport)** for PR 22504 at commit [`6e14137`](https://github.com/apache/spark/commit/6e141373910df8c159429c367a8e937ad71eccbe).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/97575/
Test FAILed.

ankuriitg committed:
Minor change to logic of cleaning logs

SparkQA commented:
**[Test build #97617 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97617/testReport)** for PR 22504 at commit [`a3b6550`](https://github.com/apache/spark/commit/a3b6550704031c2717364f582190be9661c4c934).

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -116,6 +117,13 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
   private[history] val fs: FileSystem = new Path(logDir).getFileSystem(hadoopConf)
 
+  private val driverLogFs: Option[FileSystem] =
+    if (conf.get(DRIVER_LOG_DFS_DIR).isDefined) {
```
Use `conf.get(DRIVER_LOG_DFS_DIR).map`.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -274,11 +282,20 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
           getRunner(() => cleanLogs()), 0, CLEAN_INTERVAL_S, TimeUnit.SECONDS)
       }
+
+      driverLogFs.foreach { _ =>
```
Here it's clearer to use `if (driverLogFs.isDefined && ...)`.

But really this is a bit confusing and I think my previous suggestion is clearer: don't keep `driverLogFs` in a field, just get it when running the cleaner task. And do this initialization based on the config entries, not `driverLogFs`.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -798,14 +815,52 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
+    while (appDirs.hasNext()) {
+      val appDirStatus = appDirs.next()
+      if (appDirStatus.isDirectory()) {
+        val logFiles = driverLogFs.get.listStatus(appDirStatus.getPath())
```
Not sure how I missed this before, but why are you storing log files in a subdirectory? What's the advantage? You'll have a ton of directories with a single log file. Instead, just write the log files directly into the parent directory... that also makes this code simpler / faster (less RPCs).

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -798,14 +815,52 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
+        var deleteDir = true
+        logFiles.foreach { f =>
+          try {
+            val info = listing.read(classOf[LogInfo], f.getPath().toString())
```
This block probably deserves a comment.

vanzin replied:
Oh, an you probably want unit tests for this new code.

vanzin commented on core/src/main/scala/org/apache/spark/util/Utils.scala:
```diff
@@ -2863,6 +2870,14 @@ private[spark] object Utils extends Logging {
     if (str == null) 0 else str.length + fullWidthRegex.findAllIn(str).size
   }
+
+  private[spark] def sanitizeDirName(str: String): String = {
```
This class is already `private[spark]` so the modifier is redundant.

vanzin commented on core/src/test/scala/org/apache/spark/util/logging/DriverLoggerSuite.scala:
```diff
@@ -0,0 +1,79 @@
+    val dfsDir = FileUtils.getFile(sc.getConf.get(DRIVER_LOG_DFS_DIR).get, app_id)
+    assert(dfsDir.exists())
+    val dfsFiles = dfsDir.listFiles()
+    dfsFiles.exists{ f => f.getName().equals(DriverLogger.DRIVER_LOG_FILE) && f.length() > 0 }
```
space before `{`

You also need to assert the condition...

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/97617/
Test PASSed.

ankuriitg committed:
Incorporate review feedback: Part 6

1. Minor refactoring
2. Added a unit test case
3. Driver log files now reside directly inside the configured Spark driver log dfs dir, names starting with appId

SparkQA commented:
**[Test build #97624 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97624/testReport)** for PR 22504 at commit [`0bd33f6`](https://github.com/apache/spark/commit/0bd33f6b2fccc45fcd99de735b8a25465aedb325).

SparkQA commented:
**[Test build #97624 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97624/testReport)** for PR 22504 at commit [`0bd33f6`](https://github.com/apache/spark/commit/0bd33f6b2fccc45fcd99de735b8a25465aedb325).
 * This patch **fails PySpark pip packaging tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/97624/
Test FAILed.

tgravescs commented on docs/configuration.md:
```diff
@@ -266,6 +266,41 @@ of the most common options to set are:
+    is true. Within this base directory, Spark creates a sub-directory for each application, and logs the driver
+    logs specific to the application in this directory. Users may want to set this to a unified location like an
+    HDFS directory so driver log files can be persisted for later usage. This directory should allow any Spark
+    user to read/write files and the Spark History Server user to delete files. Additionally, older logs from
```
we should add something about this to the security doc with specific information on permissions, like for event logging: https://spark.apache.org/docs/latest/security.html

ankuriitg committed:
Added driver log dir permission management information to security doc

SparkQA commented:
**[Test build #97934 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97934/testReport)** for PR 22504 at commit [`a620bcc`](https://github.com/apache/spark/commit/a620bcc947111ad9530841fd5de20c86c477226e).

SparkQA commented:
**[Test build #97934 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97934/testReport)** for PR 22504 at commit [`a620bcc`](https://github.com/apache/spark/commit/a620bcc947111ad9530841fd5de20c86c477226e).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/97934/
Test FAILed.

SparkQA commented:
**[Test build #4394 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4394/testReport)** for PR 22504 at commit [`a620bcc`](https://github.com/apache/spark/commit/a620bcc947111ad9530841fd5de20c86c477226e).

squito commented:
known test failure addressed here: https://github.com/apache/spark/pull/22808

SparkQA commented:
**[Test build #4394 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4394/testReport)** for PR 22504 at commit [`a620bcc`](https://github.com/apache/spark/commit/a620bcc947111ad9530841fd5de20c86c477226e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -274,11 +275,20 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
           getRunner(() => cleanLogs()), 0, CLEAN_INTERVAL_S, TimeUnit.SECONDS)
       }
+
+      conf.get(DRIVER_LOG_DFS_DIR).foreach { _ =>
```
`if (conf.contains(DRIVER_LOG_DFS_DIR) && conf.get(DRIVER_LOG_CLEANER_ENABLED))`

Don't use `foreach` as a boolean check.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -796,16 +806,57 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
     stale.foreach { log =>
-      if (log.appId.isEmpty) {
+      if (log.appId.isEmpty &&
+          (!conf.get(DRIVER_LOG_DFS_DIR).isDefined ||
```
This check is kinda awkward. How about a new property in `LogInfo` with the type of the log?

You could then also just filter those out above before `.toList`, since the `clearBlacklist` call is unrelated to the logs you're adding.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -796,16 +806,57 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
+    val currentTime = clock.getTimeMillis()
+    val maxTime = currentTime - conf.get(MAX_DRIVER_LOG_AGE_S) * 1000
+    val logFiles = driverLogFs.listLocatedStatus(new Path(driverLogDir))
+    while (logFiles.hasNext()) {
```
One issue here is that since you're basing this on the file system's contents, if these files are deleted outside of the SHS then you'll accumulate `LogInfo` entries for those deleted entries.

The event log cleaner avoids that by basing this logic on the SHS's view of the file system, although I don't know if that same logic can be applied here.

vanzin commented on core/src/main/scala/org/apache/spark/util/Utils.scala:
```diff
@@ -2328,7 +2328,14 @@ private[spark] object Utils extends Logging {
+    val rootLogger = org.apache.log4j.Logger.getRootLogger()
+    rootLogger.setLevel(l)
+    rootLogger.getAllAppenders().asScala.foreach { tmp =>
+      tmp match {
```
You can have the cases directly in the body of the foreach.

vanzin commented on core/src/main/scala/org/apache/spark/util/logging/DriverLogger.scala:
```diff
@@ -0,0 +1,206 @@
+    private var fileSystem: FileSystem = _
+    private val dfsLogFile: String = {
+      val rootDir = conf.get(DRIVER_LOG_DFS_DIR).get
+      fileSystem = new Path(rootDir).getFileSystem(hadoopConf)
```
This is actually super weird, initializing another field in the initializer of this field.

It's looking like you should just have an `init()` function, or maybe have `rootDir` as a field so you can initialize `fileSystem` more directly...

(Sometimes I really miss Java constructors.)

vanzin commented on core/src/test/scala/org/apache/spark/deploy/history/FsHistoryProviderSuite.scala:
```diff
@@ -413,6 +417,68 @@ class FsHistoryProviderSuite extends SparkFunSuite with BeforeAndAfter with Matc
+
+    val log1 = FileUtils.getFile(testDir, "1" + DriverLogger.DRIVER_LOG_FILE_SUFFIX)
+    createEmptyFile(log1)
+    val modTime1 = System.currentTimeMillis()
```
You're mixing a manual clock with system time. Why not use `File.setLastModified` like other tests?

vanzin commented on core/src/test/scala/org/apache/spark/deploy/history/FsHistoryProviderSuite.scala:
```diff
@@ -413,6 +417,68 @@ class FsHistoryProviderSuite extends SparkFunSuite with BeforeAndAfter with Matc
+    assert(log3.exists())
+
+    // Should cleanup the second file but not the third file, as filelength changed.
+    val writer = new OutputStreamWriter(new BufferedOutputStream(new FileOutputStream(log3)))
```
`Files.write` is shorter and nicer.

ankuriitg committed:
Incorporate review feedback: Part 7

1. Use logType in LogInfo to differentiate DriverLogs with EventLogs
2. Use only manual clock in unit test
3. Delete any driver log entries for the files that are deleted externally

SparkQA commented:
**[Test build #98230 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98230/testReport)** for PR 22504 at commit [`e898d47`](https://github.com/apache/spark/commit/e898d47529e9aaa13751cc89f5d5ce4b243e430d).

SparkQA commented:
**[Test build #98230 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98230/testReport)** for PR 22504 at commit [`e898d47`](https://github.com/apache/spark/commit/e898d47529e9aaa13751cc89f5d5ce4b243e430d).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/98230/
Test FAILed.

ankuriitg committed:
Compilation Fixes

SparkQA commented:
**[Test build #98234 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98234/testReport)** for PR 22504 at commit [`426beb0`](https://github.com/apache/spark/commit/426beb0f40f69ac1d61915cb2f479300f37c8278).

ankuriitg committed:
Minor refactoring: Part 3

SparkQA commented:
**[Test build #98374 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98374/testReport)** for PR 22504 at commit [`4df08bd`](https://github.com/apache/spark/commit/4df08bd56b4cd51c4072aa026bf7f46bc574421d).

SparkQA commented:
**[Test build #98374 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98374/testReport)** for PR 22504 at commit [`4df08bd`](https://github.com/apache/spark/commit/4df08bd56b4cd51c4072aa026bf7f46bc574421d).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/98374/
Test FAILed.

SparkQA commented:
**[Test build #98401 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98401/testReport)** for PR 22504 at commit [`4df08bd`](https://github.com/apache/spark/commit/4df08bd56b4cd51c4072aa026bf7f46bc574421d).

vanzin reviewed (commented):
Mostly minor things except for the change to the `LogInfo` type which may need some tweaks.

SparkQA commented:
**[Test build #4409 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4409/testReport)** for PR 22504 at commit [`4df08bd`](https://github.com/apache/spark/commit/4df08bd56b4cd51c4072aa026bf7f46bc574421d).

squito commented:
Jenkins, add to whitelist

SparkQA commented:
**[Test build #98407 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98407/testReport)** for PR 22504 at commit [`4df08bd`](https://github.com/apache/spark/commit/4df08bd56b4cd51c4072aa026bf7f46bc574421d).

ankuriitg committed:
Incorporate review feedback: Part 8

1. Tested older LogInfo objects are compatible with newer ones by:
  a. Enabled disk storage
  b. Created event logs in the older format
  c. Updated SHS with LogInfo changes and restarted SHS
  d. Ensured SHS lists older event logs

SparkQA commented:
**[Test build #98401 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98401/testReport)** for PR 22504 at commit [`4df08bd`](https://github.com/apache/spark/commit/4df08bd56b4cd51c4072aa026bf7f46bc574421d).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/98401/
Test FAILed.

SparkQA commented:
**[Test build #98413 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98413/testReport)** for PR 22504 at commit [`70a227f`](https://github.com/apache/spark/commit/70a227fd83524f35264fac9717d07024f440d179).

SparkQA commented:
**[Test build #98407 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98407/testReport)** for PR 22504 at commit [`4df08bd`](https://github.com/apache/spark/commit/4df08bd56b4cd51c4072aa026bf7f46bc574421d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/98407/
Test PASSed.

SparkQA commented:
**[Test build #98413 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98413/testReport)** for PR 22504 at commit [`70a227f`](https://github.com/apache/spark/commit/70a227fd83524f35264fac9717d07024f440d179).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/98413/
Test PASSed.

vanzin commented on docs/configuration.md:
```diff
@@ -266,6 +266,40 @@ of the most common options to set are:
+    Users may want to set this to a unified location like an HDFS directory so driver log files can be persisted
+    for later usage. This directory should allow any Spark user to read/write files and the Spark History Server
+    user to delete files. Additionally, older logs from this directory are cleaned by
+    <a href="monitoring.html#spark-history-server-configuration-options"> Spark History Server</a>  if
```
remove space after `>`

vanzin commented on docs/configuration.md:
```diff
@@ -266,6 +266,40 @@ of the most common options to set are:
+    is true. Within this base directory, each application logs the driver logs to an application specific file.
+    Users may want to set this to a unified location like an HDFS directory so driver log files can be persisted
+    for later usage. This directory should allow any Spark user to read/write files and the Spark History Server
+    user to delete files. Additionally, older logs from this directory are cleaned by
```
...cleaned by the...

vanzin commented on docs/configuration.md:
```diff
@@ -266,6 +266,40 @@ of the most common options to set are:
+    user to delete files. Additionally, older logs from this directory are cleaned by
+    <a href="monitoring.html#spark-history-server-configuration-options"> Spark History Server</a>  if
+    <code>spark.history.fs.driverlog.cleaner.enabled</code> is true and, if they are older than max age configured
+    at <code>spark.history.fs.driverlog.cleaner.maxAge</code>.
```
s/at/by setting

vanzin commented on docs/configuration.md:
```diff
@@ -266,6 +266,40 @@ of the most common options to set are:
+    If true, spark application running in client mode will write driver logs to a persistent storage, configured
+    in <code>spark.driver.log.dfsDir</code>. If <code>spark.driver.log.dfsDir</code> is not configured, driver logs
+    will not be persisted. Additionally, enable the cleaner by setting <code>spark.history.fs.driverlog.cleaner.enabled</code>
+    to true in <a href="monitoring.html#spark-history-server-configuration-options"> Spark History Server</a>.
```
no space after `>`

vanzin commented on docs/configuration.md:
```diff
@@ -266,6 +266,40 @@ of the most common options to set are:
+  <td>%d{yy/MM/dd HH:mm:ss.SSS} %t %p %c{1}: %m%n</td>
+  <td>
+    The layout for the driver logs that are synced to <code>spark.driver.log.dfsDir</code>. If 
+    <code>spark.driver.log.persistToDfs.enabled</code> is true and this configuration is used. If this is not configured,
```
No need to mention the `enabled` option here.

vanzin commented on docs/monitoring.md:
```diff
@@ -202,6 +202,28 @@ Security options for the Spark History Server are covered more detail in the
+    <td>spark.history.fs.driverlog.cleaner.interval</td>
+    <td><code>spark.history.fs.cleaner.interval</code></td>
+    <td>
+      How often the filesystem driver log history cleaner checks for files to delete.
```
driver log cleaner

vanzin reviewed (commented):
Looks good, but you need to merge with the current master.

ankuriitg committed:
Minor updates to docs

SparkQA commented:
**[Test build #98562 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98562/testReport)** for PR 22504 at commit [`81ed77b`](https://github.com/apache/spark/commit/81ed77b51c305ce4c48135150604a7607d1e44be).

SparkQA commented:
**[Test build #98562 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98562/testReport)** for PR 22504 at commit [`81ed77b`](https://github.com/apache/spark/commit/81ed77b51c305ce4c48135150604a7607d1e44be).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/98562/
Test FAILed.

ankuriitg committed:
Minor scalastyle fix

SparkQA commented:
**[Test build #98564 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98564/testReport)** for PR 22504 at commit [`616bd01`](https://github.com/apache/spark/commit/616bd01b5b23dc0ce0b50e95b7fd6684b1f7e9d5).

SparkQA commented:
**[Test build #98564 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98564/testReport)** for PR 22504 at commit [`616bd01`](https://github.com/apache/spark/commit/616bd01b5b23dc0ce0b50e95b7fd6684b1f7e9d5).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/98564/
Test FAILed.

ankuriitg commented:
retest this please

SparkQA commented:
**[Test build #98570 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98570/testReport)** for PR 22504 at commit [`616bd01`](https://github.com/apache/spark/commit/616bd01b5b23dc0ce0b50e95b7fd6684b1f7e9d5).

SparkQA commented:
**[Test build #98570 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98570/testReport)** for PR 22504 at commit [`616bd01`](https://github.com/apache/spark/commit/616bd01b5b23dc0ce0b50e95b7fd6684b1f7e9d5).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/98570/
Test PASSed.

vanzin commented:
retest this please

SparkQA commented:
**[Test build #98799 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98799/testReport)** for PR 22504 at commit [`616bd01`](https://github.com/apache/spark/commit/616bd01b5b23dc0ce0b50e95b7fd6684b1f7e9d5).

SparkQA commented:
**[Test build #98805 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98805/testReport)** for PR 22504 at commit [`616bd01`](https://github.com/apache/spark/commit/616bd01b5b23dc0ce0b50e95b7fd6684b1f7e9d5).

vanzin commented:
Merging to master.
